presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1262,13 +1262,23 @@ private void analyzeHaving(QuerySpecification node, Scope scope)
                     orderByExpression = outputExpressions.get(field);
                 }
                 else {
+                    // Analyze the original expression using a synthetic scope (which delegates to the source scope for any missing name)
+                    // to catch any semantic errors (due to type mismatch, etc)
+                    Scope synthetic = Scope.builder()
+                            .withParent(sourceScope)
+                            .markQueryBoundary() // this is needed because of how the field resolution walks scopes
+                            .withRelationType(outputScope.getRelationType())
+                            .build();
+
+                    analyzeExpression(expression, synthetic);
```

Comment:
Any semantic errors should be caught here, so errors that show up afterwards should be considered bugs.